### ISSUE TITLE
OWASP Coreruleset Integration

### ISSUE THREAD

**Issue by elkinaguas** [jcchavezs/coraza-http-wasm-traefik#7]
Hello,

I was thinking that it would be a good idea to integrate the OWASP coreruleset into this plugin. I guess this would mean translating all the rules into directives in yaml format.

Best,
Elkin

**Comment by jcchavezs** [comment]
CRS is included in this plugin (see https://github.com/jcchavezs/coraza-http-wasm/blob/main/main.go#L123).

You can include them using `Include @owasp_crs/**.conf`

**Comment by elkinaguas** [comment]
Noted, thank you.

**Comment by elkinaguas** [comment]
Hello again @jcchavezs 

Sorry if the question is very basic, but I'm new with go. I got this pluging working with a local instance (binary installation) of Traefik. I tried including `Include @owasp_crs/**.conf` in the directives of the the pluging, having the code as follows (entrypoint listening on port 80):

```yaml
---
http:
  # Add the router
  routers:
    router0:
      entryPoints:
      - web
      middlewares:
      - waf
      service: service-foo
      rule: Host(`hola.com`)

  # Add the middleware
  middlewares:
    waf:
      plugin:
        coraza:
          directives:
            - SecRuleEngine On
            - SecDebugLog /dev/stdout
            - SecDebugLogLevel 9
            - SecRule REQUEST_URI "@streq /admin" "id:101,phase:1,log,deny,status:403"
            - Include @owasp_crs/**.conf

  # Add the service
  services:
    service-foo:
      loadBalancer:
        servers:
        - url: http://0.0.0.0:8000/
        passHostHeader: false
```
However, when I do a `curl -v -H "Host: hola.com" 0.0.0.0:80` I get this:
```bash
*   Trying 0.0.0.0:80...
* TCP_NODELAY set
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host: hola.com
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 500 Internal Server Error
< Date: Mon, 22 Apr 2024 16:39:35 GMT
< Content-Length: 0
< 
* Connection #0 to host 0.0.0.0 left intact
```
And these are the last lines of the Traefik log file:
```bash
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Finished rule evaluation tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980041 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Evaluating rule tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980042 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Expanding arguments for rule tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980042 variable="TX" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Finished rule evaluation tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980042 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Evaluating rule tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Expanding arguments for rule tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 variable="TX" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Matching rule tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 variable_name="TX" key="detection_anomaly_score" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Evaluating action tx_id="TxltFxKAmxZoTKpkvKY" action="nolog" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Evaluating operator: MATCH tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 variable="TX" operator_function="@eq" operator_data="0" arg="tx.detection_inbound_anomaly_score" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Executing disruptive action for rule tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 action="pass" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Evaluating flow action for rule tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 action="skipafter" phase=5 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Starting secmarker tx_id="TxltFxKAmxZoTKpkvKY" value="END-REPORTING" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Rule matched tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Finished rule evaluation tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980043 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980044 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980045 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980046 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980047 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980048 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980049 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980050 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980051 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=0 skip_after="END-REPORTING" secmarker="LOG-REPORTING" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Skipping rule because of SkipAfter tx_id="TxltFxKAmxZoTKpkvKY" rule_id=980170 skip_after="END-REPORTING" secmarker="" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Evaluating rule tx_id="TxltFxKAmxZoTKpkvKY" rule_ref="@owasp_crs/RESPONSE-980-CORRELATION.conf#L7305" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Forcing rule to match tx_id="TxltFxKAmxZoTKpkvKY" rule_ref="@owasp_crs/RESPONSE-980-CORRELATION.conf#L7305" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Finished rule evaluation tx_id="TxltFxKAmxZoTKpkvKY" rule_ref="@owasp_crs/RESPONSE-980-CORRELATION.conf#L7305" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Finished phase tx_id="TxltFxKAmxZoTKpkvKY" phase=5 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Transaction marked for audit logging tx_id="TxltFxKAmxZoTKpkvKY" entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
2024-04-22T18:39:35+02:00 DBG github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Transaction finished tx_id="TxltFxKAmxZoTKpkvKY" is_interrupted=true status=500 rule_id=901001 entryPointName=web middlewareName=waf@file middlewareType=wasm routerName=router0@file
```
It looks to me that Traefik is evaluating the owasp rules or at least loading them somehow, but I'm not sure why I'm getting a status 500.

**Comment by jcchavezs** [comment]
This comes from

```

2024-04-22T18:39:35+02:00 DBG
github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > Transaction
finished tx_id="TxltFxKAmxZoTKpkvKY" is_interrupted=true status=500
rule_id=901001 entryPointName=web ***@***.***
middlewareType=wasm ***@***.***

```


You need to setup CRS by doing `Include @crs-setup.conf.example` and also
load coraza recommended config by doing ***@***.***` and then
load all CRS rules.

José Carlos Chávez


man. 22. apr. 2024 kl. 18:48 skrev Elkin Aguas ***@***.***>:

**Comment by elkinaguas** [comment]
I confirm it's working now, adding the `Include @crs-setup.conf.example` fixed the problem. Thanks again.

**Comment by dwrightpulse** [comment]
Thanks, working for me too.  Didn't understand from the guides etc that the CRS files were embedded as part of the solution, and was hence trying to use Traefik's WASM mounts to give this plugin access to the CRS files (see https://github.com/jcchavezs/coraza-http-wasm-traefik/issues/12).

Remaining question then is which version of the CRS ruleset is included in the 0.2.2 release???
